### PR TITLE
Issue 21011 contentlet as json 2

### LINKED ISSUE #21011: Persist Contentlets as Json
As a user I would like to be able to store a JSON representation of a piece of content in a database. This would provide the possibility of querying the database against the json object and having a more readable representation of the content. This is a first approach of how the json would look like: https://gist.github.com/wezell/142a905195de6e8e3aae34252cf2b362

Some important aspects to be considered for the implementation:
- A new column should be created in the `contentlet` table to store the json. This column will be optional. The datatype for each database would be:

  - Postgres: JSONB
  - Oracle: NCLOB
  - MySQL: text
  - MSSQL: NVARCHAR(MAX) 

#### CHANGES ON THIS CARD ARE ONLY FOR POSTGRES

- There must be a logic to create an immutable Contentlet object from the json column and viceversa, a logic to generate an immutable json given a Contentlet. 
- As the db column is optional, in case it is null or empty, we should keep using our current logic to create contentlets, namely, using the legacy db columns.
- So far, no upgrade tasks will be implemented to populate the json column in database.  
- To prevent potential collisions, all internal properties a piece of content has (i.e.: modDate, modUser, inode, identifier, languageId, showOnMenu, etc) should be stored inside a `dotSystemFields` node. For example: 
```
"dotSystemFields":{
        "inode":"inodeId",
        "identifier":"identifierId",
        "contentType":"contentTypeId",
        "modDate":"2021-10-01 13:17 -0.0",  
        "baseType":"fileAsset",
        "showOnMenu":false,
        "modUser":"dotcms.org.1",
        "sortOrder":0,
        "languageId":1,
        "editState":"disabledWysiwyg",
        "friendlyName":"this is a friendly name",
         //system level tags (all content can have tags)
        "tags":["tag1", "tag2"]
    },
```
- Related content will be stored as an array of identifiers

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
This adds storing contentlet as JSON capabilities.
A New column was added. The upsert command was modified to support the new column plus it handles the special formating required by postgres.
Also, the way we return keyValues was modified so we always get a Map and not String.

### CODE REVIEW

**Review by nollymar — CHANGES_REQUESTED** [review]
I left some comments

**Review by nollymar — APPROVED** [review]

**Review by spbolton — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPI.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package com.dotcms.content.business;
+
+import com.dotmarketing.db.DbConnectionFactory;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotSecurityException;
+import com.dotmarketing.util.Config;
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+/**
+ *
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPI.java:10]: Doc

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    }
+
+    private final Lazy<ObjectMapper> objectMapper = Lazy.of(() -> {
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.enable(SerializationFeature.INDENT_OUTPUT);
+        objectMapper.registerModule(new Jdk8Module());
+        objectMapper.registerModule(new GuavaModule());
+        objectMapper.registerModule(new JavaTimeModule());
+        objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+        return objectMapper;
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:408]: We already have an one for the Rest endpoints, but I guess Fab has a reason to config their own one

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 3]
```diff
@@ -1941,6 +1950,35 @@ protected Contentlet save(Contentlet contentlet, String existingInode)
...
+            if (null != map) {
+                try {
+
+                    if (UtilMethods.isNotSet((String) map.get("inode")) && UtilMethods
+                            .isSet(inode)) {
+                        map.put("inode", inode);
+                    }
+
+                    final String asJson = contentletJsonAPI.toJson(new Contentlet(map));
+                    Logger.info(ESContentletAPIImpl.class, asJson);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1967]: is that ok, may be so noisy

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 4]
```diff
@@ -2091,6 +2136,20 @@ private void addDynamicFields(final List<Object> upsertValues, final Map<String,
         }
     }
 
+    private void emptyDynamicFields(final List<Object> upsertValues,  final String prefix){
+        Object defaultValue = null;
+        if (prefix.equals("integer") || prefix.equals("float")){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:2141]: constants should be in the left

**Review by jdotcms — APPROVED** [review]

**Pull request merged**